server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -303,17 +310,130 @@ private CoordinationDiagnosticsResult diagnoseOnHaveNotSeenMasterRecently(Master
+                result = new CoordinationDiagnosticsResult(
+                    CoordinationDiagnosticsStatus.RED,
+                    String.format(
+                        Locale.ROOT,
+                        "No master node observed in the last %s, and the master eligible nodes are unable to form a quorum",
+                        nodeHasMasterLookupTimeframe
+                    ),
+                    getDetails(explain, localMasterHistory, null, coordinator.getClusterFormationState().getDescription())
+                );
+            } else {
+                result = new CoordinationDiagnosticsResult(
+                    CoordinationDiagnosticsStatus.RED,
+                    String.format(
+                        Locale.ROOT,
+                        "No master node observed in the last %s, and the cause has not been determined.",
+                        nodeHasMasterLookupTimeframe
+                    ),
+                    getDetails(explain, localMasterHistory, null, coordinator.getClusterFormationState().getDescription())
+                );
+            }
         }
         return result;
     }
 
+    /**
+     * This method checks whether each master eligible node has discovered each of the other master eligible nodes. If any node in
+     * nodeToClusterFormationStateMap to discover any other node in masterEligibleNodes, then this method returns true. If no discovery
+     * problems are detected, this method returns false.
+     * @param masterEligibleNodes The collection of all master eligible nodes
+     * @param nodeToClusterFormationStateMap A map of each master node to its ClusterFormationState
+     * @return true if there are discovery problems, false otherwise
+     */
+    private boolean hasDiscoveryProblems(
+        Collection<DiscoveryNode> masterEligibleNodes,
+        Map<DiscoveryNode, ClusterFormationFailureHelper.ClusterFormationState> nodeToClusterFormationStateMap
+    ) {
+        Map<DiscoveryNode, Collection<DiscoveryNode>> nodesNotDiscoveredMap = new HashMap<>();
+        for (Map.Entry<DiscoveryNode, ClusterFormationFailureHelper.ClusterFormationState> entry : nodeToClusterFormationStateMap
+            .entrySet()) {
+            Set<DiscoveryNode> foundPeersOnNode = new HashSet<>(entry.getValue().foundPeers());
+            if (foundPeersOnNode.containsAll(masterEligibleNodes) == false) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * This method checks that each master eligible node in the quorum thinks that it can form a quorum. If there are nodes that report a
+     * problem forming a quorum, this method returns true
+     * @param nodeToClusterFormationStateMap A map of each master node to its ClusterFormationState
+     * @return True if any nodes in nodeToClusterFormationStateMap report a problem forming a quorum, false otherwise.
+     */
+    private boolean hasQuorumProblems(
```

Comment:
Same as above - would it be useful to log the problems we discover? Or return them?

Reply:
Yeah at some point we'll be putting them into the details section of the response. For now I'll log them.

Reply:
Just to confirm - this means they don't currently add any new information compared to what we provide in the generic `cluster_formation.description` field. Is that correct?

Reply:
Right. That was the conclusion to the discussion on the document I shared about this:
```
W.r.t. structure I don’t think we have a clear indication as to how the details section of the `master_is_stable` indicator is going to be used for now, so I’d
suggest we keep the `ClusterFormationState#description` as the only field in the `master_is_stable` details field for now and add structure at 
a later phase. 

Once the health API is used in the diagnostics bundle (very soon) we’ll be able to get some more engineers exposed to the `details` field and get some 
feedback about the needs and shortcomings.
```